includes/class-micropub-authorize.php

```diff
@@ -0,0 +1,218 @@
+<?php
+
+/* For debugging purposes this will bypass Micropub authentication
+ * in favor of WordPress authentication
+ * Using this to test querying(q=) parameters quickly
+ */
+// Allows for a custom Authentication and Token Endpoint
+if ( ! defined( 'MICROPUB_AUTHENTICATION_ENDPOINT' ) ) {
+	define( 'MICROPUB_AUTHENTICATION_ENDPOINT', 'https://indieauth.com/auth' );
+}
+if ( ! defined( 'MICROPUB_TOKEN_ENDPOINT' ) ) {
+	define( 'MICROPUB_TOKEN_ENDPOINT', 'https://tokens.indieauth.com/token' );
+}
+
+add_action( 'plugins_loaded', array( 'Micropub_Authorize', 'init' ) );
+
+/**
+ * Micropub IndieAuth Authorization Class
+ */
+class Micropub_Authorize {
+
+	// associative array, populated by authorize().
+	protected static $micropub_auth_response;
+
+	// Array of Scopes
+	protected static $scopes;
+
+	/**
+	 * Initialize the plugin.
+	 */
+	public static function init() {
+		$cls = get_called_class();
+
+		add_action( 'wp_head', array( $cls, 'html_header' ), 99 );
+		add_action( 'send_headers', array( $cls, 'http_header' ) );
+		add_filter( 'host_meta', array( $cls, 'jrd_links' ) );
+		add_filter( 'webfinger_user_data', array( $cls, 'jrd_links' ) );
+		// The WordPress IndieAuth plugin uses priority 30
+		add_filter( 'determine_current_user', array( $cls, 'determine_current_user' ), 31 );
+	}
+
+	public static function indieauth_scopes( $scopes ) {
+		return static::$scopes;
+	}
+	public static function indieauth_response( $response ) {
+		return static::$micropub_auth_response;
+	}
+
+	public static function header( $header, $value ) {
+			header( $header . ': ' . $value, false );
+	}
+
+	public static function http_header() {
+			static::header( 'Link', '<' . get_option( 'indieauth_authorization_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT ) . '>; rel="authorization_endpoint"' );
+			static::header( 'Link', '<' . get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ) . '>; rel="token_endpoint"' );
+	}
+	public static function html_header() {
+			printf( '<link rel="authorization_endpoint" href="%s" />' . PHP_EOL, get_option( 'indieauth_authorization_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT ) );
+			printf( '<link rel="token_endpoint" href="%s" />' . PHP_EOL, get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ) );
+	}
+
+	public static function jrd_links( $array ) {
+		$array['links'][] = array(
+			'rel'  => 'authorization_endpoint',
+			'href' => get_option( 'indieauth_authorization_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT ),
+		);
+		$array['links'][] = array(
+			'rel'  => 'token_endpoint',
+			'href' => get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ),
+		);
+
+		return $array;
+	}
+
+	public static function get( $array, $key, $default = array() ) {
+		if ( is_array( $array ) ) {
+			return isset( $array[ $key ] ) ? $array[ $key ] : $default;
+		}
+		return $default;
+	}
+
+	/**
+	 * Get the authorization header
```

Comment:
this is so great! thank you for adding it! it's probably been the single biggest pain point for users for a while now. did you get the code from @zegnat?

it's probably worth a unit test. mind adding one? and also updating the docs, specifically the _Frequently Asked Questions_ section in the readme?

Reply:
This is something I put together after reading dozens of AUTHORIZATION header error messages in a Google Search. It tries the $_SERVER first before looking for getallheaders, which is an Apache only function. It isn't perfect though. I use a similar function in the IndieAuth plugin.

Reply:
As for unit tests, I'd like to build unit tests for this and the IndieAuth plugin, as they are both acting as clients and can both use the same tests, or variants thereof. I actually want to bring some of the other improvements over.

Reply:
I have another PR planned to enhance the user_url functions with what I learned in the other plugin and will be submitting it with the additional unit tests, if you feel comfortable merging as I have the start of the unit tests in that code.

Reply:
> I'd like to build unit tests for this and the IndieAuth plugin

i like that instinct toward reuse! sharing a set of unit tests across multiple different codebases is very difficult at best, and probaby impossible in most cases, like this. (i'm not sure i've _ever_ seen it done.) we should plan to just write unit tests for this plugin here, like normal.

(compliance test suites like https://micropub.rocks/ are great, of course, and obviously can be used with mutliple codebases! they're applications themselves though, not unit tests. unit tests are different.)

Reply:
also we should definitely still ask @zegnat to look at this (without blocking the PR) and maybe contribute his work too! he did a great deep dive on this a bit ago.

..aha, here: https://gist.github.com/Zegnat/f4192c3cce05d847af3547eea80e383c

Reply:
Both plugins need different unit tests, but the ones for testing auth headers would be the same. Either way, I'd like to get @Zegnat 's opinion, and maybe incorporate a version of the test code that would allow checking headers in one-click.

Reply:
Looks like my name has been said three times, so it is time for me to show up. This seems to be as complete as you are going to get it, and very close to what I am [doing in Mintoken](https://github.com/Zegnat/php-mintoken/blob/f93b888c4c56dd5faa129b6c0ccae90b7b00132f/endpoint.php#L196-L202). In fact, I think I am going to adapt the `REDIRECT_HTTP_AUTHORIZATION` into Mintoken as well! That’s a great idea!

The only real choices you get to make in this implementation are:

1. In what order do you check? (`$_SERVER` vs. server-specific functions.) I think checking `$_SERVER` first is valid and saves you a function check on all non-apache servers, though such micro-optimisations are probably worthless in the grand scheme of a WordPress extension.
   
2. Do you use `getallheaders`, `apache_request_headers`, or both, for the server-specific function? For the diagnostics tool I check both, but so far have not run into any servers that supported one without the other. That’s why I only use one in Mintoken.
   
   I choose to use `apache_request_headers` because it clearly illustrates that it isn’t going to work anywhere else. But that is a personal preference thing more than anything else.

:+1: from me.

Reply:
getallheaders is an Apache  function, if I remember correctly, is an alias for apache_request_headers.(https://secure.php.net/manual/en/function.getallheaders.php). That is why, if it isn't defined, the built-in version takes over, which effectively does the same thing. The reasoning for checking $_SERVER first is more that, is that PHP itself can continue to improve how it loads the server environment variable. It isn't about the optimization so much as allowing other code that may get upgraded in future to try.

Reply:
Have you seen/tried: https://developer.wordpress.org/reference/classes/wp_rest_request/get_header/

Reply:
> Have you seen/tried: https://developer.wordpress.org/reference/classes/wp_rest_request/get_header/

Looking at that it relies 100% on an array that it has previously been fed through `set_headers`. This happens [in the WP_REST_Server class](https://github.com/WordPress/WordPress/blob/4c2fcf36a3193a7e63a82e5ff459f2587c69a859/wp-includes/rest-api/class-wp-rest-server.php#L309), but relies only on `$_SERVER` (no Apache fallback) and uses [a standard looping method](https://github.com/WordPress/WordPress/blob/4c2fcf36a3193a7e63a82e5ff459f2587c69a859/wp-includes/rest-api/class-wp-rest-server.php#L1315-L1342) to only extract keys starting with `HTTP_` (missing the `REDIRECT_HTTP_AUTHORIZATION` key).